### PR TITLE
Feature/meaningful words finder

### PR DESCRIPTION
**Pull request by alon1samuel**
How to find the most meaningful words in a text with an example on Hamlet text. 

1 file, with functions inside

### CODE REVIEW

File: find_meaningful_words.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import re
+from collections import Counter
+#
+WORDS_PLACE_IN_FREQUENCY_LIST = 0
+FREQUENCY_PLACE_IN_FREQUENCY_LIST = 1
+
+# TODO: Add documentation to each function. 
+# TODO: Add tests to each function
```
Comment (liordon) [inline find_meaningful_words.py]: you have failed the CR right about now 😛

File: find_meaningful_words.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+def read_file(file_path):
+    with open(file_path, "r") as f:
+        file_string = f.read()
+    return file_string
+
+
+def get_words_only_in_string(file_string):
+    list_just_words = re.findall(r'\w+', file_string)
+    list_words_small_letters = [x.lower() for x in list_just_words]
```
Comment (liordon) [inline find_meaningful_words.py]: this is kind of nit-picking, but this line performs hidden functionality. you do not declare converting the text to lower-case.

File: find_meaningful_words.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    return list_words_small_letters
+
+
+def get_words_in_file(file_path):
```
Comment (liordon) [inline find_meaningful_words.py]: this function *counts* the words in said file. it most definitely does not get them.

File: find_meaningful_words.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    file_string = read_file(file_path)
+    only_words_list = get_words_only_in_string(file_string)
+    return len(only_words_list)
+
+
+def sort_frequencies_list(frequencies_list):
+    sorted_frequencies = sorted(
+        frequencies_list, key=lambda x: x[1], reverse=True)
```
Comment (liordon) [inline find_meaningful_words.py]: should that 1 be `FREQUENCY_PLACE_IN_FREQUENCY_LIST `?

also, you don't need that variable since this function is a one-liner.

File: find_meaningful_words.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    words_counter = dict(Counter(words_list))
+    counted_words = sort_frequencies_list(words_counter.items())
+    return counted_words
+
+
+def get_only_common_words(words_frequencies_dict, number_of_most_common):
+    return [x[WORDS_PLACE_IN_FREQUENCY_LIST] for x in words_frequencies_dict[:number_of_most_common]]
+
+
+def get_most_common_words_in_play(file_path, number_of_most_common):
```
Comment (liordon) [inline find_meaningful_words.py]: play? what play?

File: find_meaningful_words.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+    file_string = read_file(file_path)
+    only_words_list = get_words_only_in_string(file_string)
+    words_frequencies_dict = count_words_in_list(only_words_list)
```
Comment (liordon) [inline find_meaningful_words.py]: it isn't a dictionary, though, is it? 😜

File: find_meaningful_words.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    most_common_words = get_only_common_words(
+        words_frequencies_dict, number_of_most_common)
+    return most_common_words
+
+
+def turn_frequencies_into_words(frequencies_list):
+    return [x[0] for x in frequencies_list]
```
Comment (liordon) [inline find_meaningful_words.py]: `WORDS_PLACE_IN_FREQUENCY_LIST`?

File: find_meaningful_words.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+
+def create_whole_frequencies_list(base_words, reference_words, reference_frequencies):
+    base_words_not_in_ref = set(base_words) ^ set(reference_words)
+    ref_freq_added_base_words = reference_frequencies + \
```
Comment (liordon) [inline find_meaningful_words.py]: should work on that variable name. make it more friendly.

File: find_meaningful_words.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    most_common_words = get_only_common_words(
+        words_frequencies_dict, number_of_most_common)
+    return most_common_words
+
+
+def turn_frequencies_into_words(frequencies_list):
+    return [x[0] for x in frequencies_list]
+
+
+def create_whole_frequencies_list(base_words, reference_words, reference_frequencies):
```
Comment (liordon) [inline find_meaningful_words.py]: what does `whole` mean in this context?

Comment (liordon) [reply find_meaningful_words.py]: after reading the function I can definitely say I have no clue what's going on 😅

File: find_meaningful_words.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    base_words_not_in_ref = set(base_words) ^ set(reference_words)
+    ref_freq_added_base_words = reference_frequencies + \
+        [(x, 1) for x in base_words_not_in_ref]
+    ref_words_whole = turn_frequencies_into_words(ref_freq_added_base_words)
+    return ref_words_whole, ref_freq_added_base_words
+
+
+def calculate_relative_from_frequencies(base_words, base_freq, ref_words_whole, ref_freq_added_base_words):
```
Comment (liordon) [inline find_meaningful_words.py]: whaaaaa?

File: find_meaningful_words.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+def get_relative_frequencies(base_freq, ref_freq):
+    base_words = turn_frequencies_into_words(base_freq)
+    ref_words = turn_frequencies_into_words(ref_freq)
+    ref_words_whole, ref_freq_added_base_words = create_whole_frequencies_list(
+        base_words, ref_words, ref_freq)
+    base_relative = calculate_relative_from_frequencies(
+        base_words, base_freq, ref_words_whole, ref_freq_added_base_words)
+
+    return base_relative
+    pass
```
Comment (liordon) [inline find_meaningful_words.py]: oopsie

File: find_meaningful_words.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+def calculate_relative_from_frequencies(base_words, base_freq, ref_words_whole, ref_freq_added_base_words):
+
+    ref_ind = [ref_words_whole.index(x) for x in base_words]
+    base_relative = [(x[WORDS_PLACE_IN_FREQUENCY_LIST], x[FREQUENCY_PLACE_IN_FREQUENCY_LIST] /
+                      ref_freq_added_base_words[ref_ind[ind]][FREQUENCY_PLACE_IN_FREQUENCY_LIST]) 
+                      for ind, x in enumerate(base_freq)]
+    return base_relative
+
+
+def get_relative_frequencies(base_freq, ref_freq):
```
Comment (liordon) [inline find_meaningful_words.py]: at this point, I see a function that I don't understand comprised of smaller functions that I don't understand.
what are these relative frequencies?

File: find_meaningful_words.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    ref_words_whole, ref_freq_added_base_words = create_whole_frequencies_list(
+        base_words, ref_words, ref_freq)
+    base_relative = calculate_relative_from_frequencies(
+        base_words, base_freq, ref_words_whole, ref_freq_added_base_words)
+
+    return base_relative
+    pass
+
+
+def find_meaningful_words_in_file_from_other(base_path, referance_path, number_of_words):
```
Comment (liordon) [inline find_meaningful_words.py]: from other what?

File: find_meaningful_words.py (hunk 10 above)
Comment (liordon) [reply find_meaningful_words.py]: relative to what? who's relative?

File: find_meaningful_words.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    most_meaningful_words = get_only_common_words(
+        base_relative_frequencies, number_of_words)
+    return most_meaningful_words
+
+
+def main():
+    hamlet_path = "hamlet.txt"
+    # Find how many words are there in hamlet txt file.
+    hamlet_words_number = get_words_in_file(hamlet_path)
+    print(f"There are {hamlet_words_number} words in hamlet plat.")
```
Comment (liordon) [inline find_meaningful_words.py]: plat?

File: find_meaningful_words.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    # Find the 30 most common words.
+    how_many_common_words = 30
+    most_common_words = get_most_common_words_in_play(
+        hamlet_path, how_many_common_words)
+    print(
+        f"The {how_many_common_words} most common words in hamlet are - {most_common_words}")
+
+    # Having macbeth txt, could you find the most meaningful words in the hamlet file.
+    macbeth_path = "macbeth.txt"
+    hamlet_meaningful_words = find_meaningful_words_in_file_from_other(
```
Comment (liordon) [inline find_meaningful_words.py]: hamlet?

File: find_meaningful_words.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        hamlet_path, macbeth_path, how_many_common_words)
+    print(
+        f"These are the most meaningful words in the hamlet play: {hamlet_meaningful_words}")
```
Comment (liordon) [inline find_meaningful_words.py]: I thought this was mcbeth 😛

File: find_meaningful_words.py (hunk 1 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Thanks! haha
I'm wondering what about :

TODO: Add documentation to each function. 
TODO: Divide function into different files. 
TODO: Add configuration file for magic numbers and strings. 

You would think of doing them also? or only the tests?

File: find_meaningful_words.py (hunk 2 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Done. I've divided these functionalities

File: find_meaningful_words.py (hunk 3 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Done. Name changed

File: find_meaningful_words.py (hunk 4 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Done, changed 1 and removed variable 'sorted_frequencies'

File: find_meaningful_words.py (hunk 5 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Done, changed to string

File: find_meaningful_words.py (hunk 13 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Haha... Changed  :)

File: find_meaningful_words.py (hunk 14 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Changed

File: find_meaningful_words.py (hunk 15 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Yes, it's hamlets meaningful words. Maybe the documentation isn't clear enough?
Or because the variables aren't there?
I've added them now for less confusion

File: find_meaningful_words.py (hunk 16 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: :laughing:

File: find_meaningful_words.py (hunk 6 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Yeah... I didn't look into it.. maybe.. :laughing:

File: find_meaningful_words.py (hunk 7 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Changed

File: find_meaningful_words.py (hunk 8 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Changed :)

File: find_meaningful_words.py (hunk 11 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Deleted.. haha

File: find_meaningful_words.py (hunk 12 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: I changed the functions tree and names. 
I was thinking I made the alg hard to understand. 
I think now it's clearer :)

File: find_meaningful_words.py (hunk 10 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Changed

File: find_meaningful_words.py (hunk 9 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: I think whole was a vague term. I changed

File: find_meaningful_words.py (hunk 1 above)
Comment (liordon) [reply find_meaningful_words.py]: just do whatever you think needs doing. since this is a small project this will not be a lot even if you choose to follow through with everything.

File: find_meaningful_words.py (hunk 15 above)
Comment (liordon) [reply find_meaningful_words.py]: I was being sarcastinc. at this point you switched from hamlet to mcbeth but the variables and outputs are still hamlet's

File: .bashrc [hunk 17]
```diff
@@ -0,0 +1 @@
+PYTHONPATH="/home/razor/Projects/alon/meaningful_words_in_txt/"
```
Comment (liordon) [inline .bashrc]: why is your bashrc inside this project? OO
this is a pretty personal preference file for your computer.

File: src/count_words.py [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+from src.utils import SaveRead
+
+
+def gets_words_list_from_string(file_string: str) -> list:
+    list_just_words = re.findall(r'\w+', file_string)
+    words_list = [x for x in list_just_words]
+    return words_list
+
+
+def convert_to_lower_letter(words_list):
```
Comment (liordon) [inline src/count_words.py]: lower case. no need to be stingy with words.

File: src/count_words.py [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+    return lower_letters_words_list
+
+
+def count_words_from_file(file_path):
+    file_string = SaveRead.read_txt_file_to_string(file_path)
+    lower_letters_words_list = get_words_only_small_letters(file_string)
+    return len(lower_letters_words_list)
+
+
+def get_words_only_small_letters(file_string: str) -> list:
```
Comment (liordon) [inline src/count_words.py]: small letters? 😛 

BTW, nice type hints 😄

File: src/find_meaningful_words.py [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+WORDS_PLACE_IN_FREQUENCY_LIST = 0
+FREQUENCY_PLACE_IN_FREQUENCY_LIST = 1
+
+# TODO: Add documentation to each function.
+# TODO: Add tests to each function
+# TODO: Divide function into different files.
+# TODO: Add configuration file for magic numbers and strings.
+
+
+def sort_words_count(counted_words_tuple_list):
```
Comment (liordon) [inline src/find_meaningful_words.py]: you can be more specific with this name. you can try to give someone the name and a sample input, and see if they guess correctly what the function should return.

File: src/find_meaningful_words.py [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    sorted_counted_words_tuple_list = sorted(
+        counted_words_tuple_list, key=lambda x: x[FREQUENCY_PLACE_IN_FREQUENCY_LIST], reverse=True)
+    return sorted_counted_words_tuple_list
+
+
+def count_words_in_list(words_list):
```
Comment (liordon) [inline src/find_meaningful_words.py]: so... this is a function that gets a mixed list like ["duck", 2, 3.14, "duck", 666, "goose"] and returns 3, right? 😜

File: src/find_meaningful_words.py [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+    counted_words_dict = dict(Counter(words_list))
+    counted_words_tuple_list = counted_words_dict.items()
+    sorted_counted_words_tuple_list = sort_words_count(
+        counted_words_tuple_list)
+    return sorted_counted_words_tuple_list
+
+
+def get_only_common_words(words_frequencies_dict, number_of_most_common):
```
Comment (liordon) [inline src/find_meaningful_words.py]: this is pretty good. I would call this `extract_most_common_words` but that's mainly a matter of preference at this point.
notice that one of your argument is called a dict, but it seems to be a list of tuples. is this a lie? 😨

File: src/find_meaningful_words.py [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+    return [x[WORDS_PLACE_IN_FREQUENCY_LIST] for x in words_frequencies_dict[:number_of_most_common]]
+
+
+def get_most_common_words_in_string(file_path, number_of_most_common):
```
Comment (liordon) [inline src/find_meaningful_words.py]: should be `get_most_common_words_in_file` since your argument is a file path

File: src/find_meaningful_words.py [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+
+
+def turn_words_count_tuple_list_to_words_only(frequencies_list):
+    return [x[WORDS_PLACE_IN_FREQUENCY_LIST] for x in frequencies_list]
+
+
+def add_missing_base_words_to_reference_words_count(base_words, reference_words, reference_frequencies):
+    base_words_not_in_ref = set(base_words) ^ set(reference_words)
+    reference_counted_words_and_base_words = reference_frequencies + \
+        [(x, 1) for x in base_words_not_in_ref]
```
Comment (liordon) [inline src/find_meaningful_words.py]: just for clarification, this inserts each reference word with the frequency of 100%?

File: .bashrc (hunk 17 above)
Comment (alon1samuel) [reply .bashrc]: I removed it... haha. 
I was testing some stuff that didn't work for me with Pytest running the tests
I didn't need it there eventually, I deleted it in the next commits

File: find_meaningful_words.py (hunk 15 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Ah no, I'm not switching to Macbeth. I'm using Macbeth to find relevant words in Hamlet. That's why the variables names are still Hamlet's

File: find_meaningful_words.py (hunk 1 above)
Comment (alon1samuel) [reply find_meaningful_words.py]: Cools... I'm kinda doing these todos to check myself. I think I will do all of them cause just having trying with the tests thought me a lot

**Comment by alon1samuel** [comment]
Liordon, Ya Ah

File: src/find_meaningful_words.py (hunk 24 above)
Comment (liordon) [reply src/find_meaningful_words.py]: come to think of it, what are these reference words exactly and why are they important?

File: src/find_meaningful_words.py [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+        reference_words_count_tuple_list)
+    base_words_with_reference_count = get_base_words_with_reference_count(
+        base_words_list, reference_words_list, reference_words_count_tuple_list)
+    base_words_relative_reference_count_tuple_list = calculate_relative_words_count_of_2_lists(
+        base_words_count_tuple_list, base_words_with_reference_count)
+
+    return base_words_relative_reference_count_tuple_list
+
+
+def find_meaningful_words_in_file_from_another_file(base_path, reference_path, number_of_words):
```
Comment (liordon) [inline src/find_meaningful_words.py]: what?

File: src/find_meaningful_words.py [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+    hamlet_path = "hamlet.txt"
+    how_many_common_words = 30
+    hamlet_most_common_words = get_most_common_words_in_string(
+        hamlet_path, how_many_common_words)
+    if print_results:
+        print(
+            f"The {how_many_common_words} most common words in hamlet are - {hamlet_most_common_words}")
+
+    # Having macbeth txt, could you find the most meaningful words in the hamlet file.
+    hamlet_path = "hamlet.txt"
```
Comment (liordon) [inline src/find_meaningful_words.py]: why is hamlet_path redefined?

File: src/utils.py [hunk 27]
```diff
@@ -0,0 +1,61 @@
+# Save a dictionary into a pickle file.
+import pickle
+import os
+import pandas as pd
+import json
+
+
+class SaveRead():
```
Comment (liordon) [inline src/utils.py:8]: you don't need this class. everything is static here anyway, so it might as well be pulled out.

File: tests/count_words_test.py [hunk 28]
```diff
@@ -0,0 +1,28 @@
+import unittest
+from src import count_words
+
+
+class TemplateTest(unittest.TestCase):
+
+    @staticmethod
+    def test_gets_words_from_string():
```
Comment (liordon) [inline tests/count_words_test.py]: oh! now this is the fun part:
the name rule for tests is that the name should tell you the spec.
in other words, when a test fails, you want to know exactly what you wanted to check and failed to do.

in this example, the test name is `test_gets_words_from_string`. who? what's checked here? if this fails, what's broken?

Comment (liordon) [reply tests/count_words_test.py]: another guidline is that each test should only check one thing. in this test I already noticed 3 checks:
1. you are checking that punctuation marks are not words or part of words
2. you are checking that sequences of letter *are* words
3. you are checking that random character sequences (~~) are not considered words

File: tests/count_words_test.py [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+        # Input data and reference
+        string_of_words_input = "Alon, guy , ~~: that are not words"
+        reference_words_list = ["Alon", "guy", "that", "are", "not", "words"]
+        # Check assertion
+        assert reference_words_list == count_words.gets_words_list_from_string(
+            string_of_words_input)
+
+    @staticmethod
+    def test_convert_to_lower_letter():
```
Comment (liordon) [inline tests/count_words_test.py]: name comment applies here as well. and this test also tests more than one thing - you tried to squeeze a different edge-case for each word in your list.

File: tests/tests_template.py [hunk 30]
```diff
@@ -0,0 +1,10 @@
+import unittest
+
+
+class TemplateTest(unittest.TestCase):
```
Comment (liordon) [inline tests/tests_template.py:4]: I guess this one is just for you

**Review by liordon — CHANGES_REQUESTED** [review]
so I've done another round of this CR. I still don't understand the reference word thing.

File: .bashrc (hunk 17 above)
Comment (liordon) [reply .bashrc]: be mindful that git saves history +-forever, so every password or secret file that you inserted and deleted in a later commit is compromised unless you took special steps to make sure otherwise.

Comment (alon1samuel) [reply .bashrc]: Thanks!! Yeah I heard that for deleting also large files you would need to delete them from history also...

File: src/count_words.py (hunk 18 above)
Comment (alon1samuel) [reply src/count_words.py]: Haha.. Nice, changed

File: src/count_words.py (hunk 19 above)
Comment (alon1samuel) [reply src/count_words.py]: Yay... :) I've thought of adding it, maybe after tests, I'll add it to all functions

Comment (alon1samuel) [reply src/count_words.py]: Small letters... yeah, must be a moment of little creativity... haha

File: src/find_meaningful_words.py (hunk 21 above)
Comment (alon1samuel) [reply src/find_meaningful_words.py]: Yeah, I think I can get more accurate. Also maybe the whole pipeline counts words, but you're right, this function just counts occurrences of elements :smile_cat:

File: src/find_meaningful_words.py (hunk 20 above)
Comment (alon1samuel) [reply src/find_meaningful_words.py]: Changed to - sort_elements_and_occurences_tuple_list_from_high_to_low

I'm kinda thinking it's too long, maybe it would be hard to be more precise on what exactly this function is doing unless I would right something long?

File: src/find_meaningful_words.py (hunk 22 above)
Comment (alon1samuel) [reply src/find_meaningful_words.py]: Thanks!! Changed both
I think I can be even more specific and maybe that can be discussed. 
Because I'm dealing with words from the context I wrote words in the function. 
But actually it words for any elements and counts tuple list 
What do you think? Would you change it ? to your experience

File: src/find_meaningful_words.py (hunk 26 above)
Comment (alon1samuel) [reply src/find_meaningful_words.py]: I think it's a matter of preference, it's redefined to help the code maintainer understand that it's being used again. 
Maybe no need?

Comment (alon1samuel) [reply src/find_meaningful_words.py]: What do you think, would you do it?

I've added it after I saw that when you looked at the third function and thought I'm not using the Hamlet file but I'm using the Macbeth file instead. So I redefined it to help have better clarity.

File: src/utils.py (hunk 27 above)
Comment (alon1samuel) [reply src/utils.py:8]: That's also a preference I would think. 
It helps encapsulate similar methods under the same class. 
But maybe no need? What would you do?

File: tests/tests_template.py (hunk 30 above)
Comment (alon1samuel) [reply tests/tests_template.py:4]: Hmmmm good question weather to commit this or not, is that what you're saying?

Comment (alon1samuel) [reply tests/tests_template.py:4]: I'm kinda thinking that it is going to help someone developing in future cases, cause it helps him write tests faster

File: tests/count_words_test.py (hunk 28 above)
Comment (alon1samuel) [reply tests/count_words_test.py]: Nice!!! Alright, 
I'll have another go on this, didn't think about it that way. 
I thought unit tests are testing basically each function.

**Comment by alon1samuel** [comment]
I've added some more detailed tests, for 1 function. 
How does it looks?

File: .bashrc (hunk 17 above)
Comment (liordon) [reply .bashrc]: yes, that's just to avoid having the cluttering your history and increasing project size.

File: src/utils.py (hunk 27 above)
Comment (liordon) [reply src/utils.py:8]: you've already collected them to the same file, so that's enough for python.
I'd do something similar to you in java, but that's just because in java I would have no choice.

Comment (liordon) [reply src/utils.py:8]: bottom line I would not use a class\object unless I need to be stateful or make subdivision within a file.
